add a bitequal() function

andre-merzky opened the pull request:
I added a bitequal() function to calculate the same as (a & b).count().

Thanks for this great library.
Björn

---

This PR was moved from ilanschnell/bitarray/pull/8.

flowerhack committed:
Minor clarity/wording changes to README

DRMacIver committed:
Handle extending self correctly

Pythons list implementation has a special case to ensure that
x.extend(x) is the same as x.extend(list(x)). i.e. it doesn't notice
changes to the size of the list during the extend.

Previously this case would not have been handled correctly by
bitarray. This changes that by checking what the size of the other
array is before we make any changes. This ensures both that this works
correctly and also (importantly) that we don't try to write past the
end of the array.

pombredanne committed:
Fixed minor doc typo.

ilanschnell committed:
add official support for Python 3.5

ilanschnell committed:
improve test output

ilanschnell committed:
Merge pull request #33 from pombredanne/patch-1

Fixed minor doc typo.

diamondman commented:
Damn test conflicts.

diamondman commented:
I love the left and right shift. But I think we should make theseoperator overloads so we can >> or <<

diamondman commented:
@udoprog @eltoder I assume equals can be sped up with SIMD. Can left/right shift or equal benefit SIMD as well? I could see issues of shifting more than the vector size.

eltoder commented:
@diamondman Yes, bitdiff() may now be slower than (a ^ b).count(), since the latter uses SIMD. Ideally we should have a SIMD version of count(), bitdiff() should use SIMD and that code as well, and this new bitequal() should reuse as much of that code as possible, instead of being a copy-paste.

Agreed that lshift() and rshift() should be hooked to python's >> and << instead. Also, these versions claim that they only work on big endian machines, and are not tested, which seems bad.

As for using SIMD for shifts: in short, it only works for constant shifts or multiples of 8. If we use `attribute vector_size`, we can see what the compiler can do for us, but I don't expect much.

diamondman commented:
I agree the big endian only is a problem. Maybe there is something woth doing with shifting with SIMD multiples of 8 and then doing the rest of shiftcount%8 shift  as a second operation. It would cause double writes, but maybe it would be faster than an arbitrary shift operation. I will try to look into this in a bit.

diamondman commented:
The code in this PR does move around the bytes first, so that is a spot for SIMD! https://github.com/diamondman/bitarray/pull/9/commits/0fd042ed49c91fff48735885a3eab1ae3862baa8#diff-1997c64b6c87f3a5197c717a65138709R2412

udoprog commented on bitarray/_bitarray.c:
```diff
@@ -2925,6 +3079,48 @@ static PyTypeObject Bitarraytype = {
+    }
+
+#define aa  ((bitarrayobject *) a)
+#define bb  ((bitarrayobject *) b)
```
Should prefer using variables here.

diamondman commented on bitarray/_bitarray.c:
```diff
@@ -2925,6 +3079,48 @@ static PyTypeObject Bitarraytype = {
 /*************************** Module functions **********************/
 
 static PyObject *
+bitequal(PyObject *self, PyObject *args)
```
Why is this a module method instead of being on the instance?

eltoder replied:
I'm guessing because bitdiff() is not a method. Arguably they both should be methods, and the names could be more clear, like count_equal() and count_different().

ilanschnell committed:
add official Python 3.6 support

ph4r05 committed:
copy_n: fast copy with memmove fixed

- fixed a problem in cases a < b when self == other. Source memory was overwritten.

ph4r05 committed:
copy_n: if n is null return immediately

ph4r05 committed:
randomised tests - more rounds, helped with debugging fast_copy in copy_n

matsjoyce committed:
Fix invalid pointer exceptions in pypy

tipabu committed:
DOC: improve description of fill()

The previous wording made it sound like the function did the opposite
of what it actually does.

yoch committed:
Efficient decode and iterdecode (use raw C tree)

yoch committed:
Adding buffer protocol support for Python 3

rrrealman commented:
Hi guys, is there still need in simple shifts? As I found here, it was not merged. So, I need this implementation and have time to commit it.

andre-merzky commented:
I think it would be a shame to let this PR deteriorate further...

rrrealman commented:
You are right. I'll create a new one so I have it already implemented.

ilanschnell committed:
update changelog

ilanschnell committed:
update changelog

ilanschnell committed:
Merge pull request #52 from tipabu/fill-doc

DOC: improve description of fill()

ilanschnell committed:
Merge pull request #28 from DRMacIver/self_extend

Handle extending self correctly

ilanschnell committed:
declare variable on the top of the function

ilanschnell committed:
update changelog; cleanup

ilanschnell committed:
update changelog

ilanschnell committed:
Merge pull request #43 from ph4r05/fast-copy

copy_n: fast copy with memmove fixed

ilanschnell committed:
we must take the size before resizing

ilanschnell committed:
add tests

ilanschnell committed:
Merge pull request #23 from flowerhack/master

Minor clarity/wording changes to README

ilanschnell committed:
add self extend test

ilanschnell committed:
simplify test

ilanschnell committed:
add release date

ilanschnell committed:
update git ignore file

ilanschnell committed:
add exception handling

ilanschnell committed:
use IOError instead

ilanschnell committed:
prepare for next release

Commit:
Get rid of warning during installation

Installing with "easy_install" causes an installation warning:
"module references __file__". zip_safe eliminates this message.

ilanschnell committed:
Merge pull request #59 from wiggin15/patch-1

Get rid of warning during installation

ilanschnell committed:
remove non-distutils option

ilanschnell committed:
update readme

ilanschnell committed:
update changelog

ilanschnell committed:
add official Python 3.7 support

ilanschnell committed:
add comment

ilanschnell committed:
Merge pull request #54 from yoch/master

Efficient decode and iterdecode

ilanschnell committed:
minor cleanup

ilanschnell committed:
update changelog

ilanschnell committed:
declare variables in beginning of function, use int

ilanschnell committed:
Py3k fixes in bloom filer example

ilanschnell committed:
make sieve example Py3k compatible

ilanschnell committed:
make ndarray example Py3k compatible

ilanschnell committed:
make compress example Py3k compatible

ilanschnell committed:
add gene sequence example

ilanschnell committed:
update changelog

ilanschnell committed:
use Python memory manager and add error handling

ilanschnell committed:
improve examples for Py3k

ilanschnell committed:
handle NULL return value

ilanschnell committed:
no reason to set prev

ilanschnell committed:
bump version, formatting

ilanschnell committed:
update docstrings in C code

ilanschnell committed:
wording

ilanschnell committed:
Merge pull request #55 from yoch/p3buffer

Adding buffer protocol support for Python 3

ilanschnell committed:
fix test for Py3k

ilanschnell committed:
Merge pull request #47 from matsjoyce/pypy

Fix invalid pointer exceptions in pypy

ilanschnell committed:
update changelog

ilanschnell committed:
fixing huffman example for Py3k

ilanschnell committed:
fix huffman example for binary files in Python 3

ilanschnell committed:
add useful output

ilanschnell committed:
update Makefile

ilanschnell committed:
ensure stable Huffman code by sorting frequency dict

ilanschnell committed:
create output format which is now consistent between Python versions - add useful print output

ilanschnell committed:
add comments

ilanschnell committed:
improve cli

ilanschnell committed:
fix segfault when decoing invalid bitarrays

ilanschnell committed:
minor style issue

ilanschnell committed:
drop Python 2.4 support

ilanschnell committed:
handle not terminated decoding properly when iterating, and add tests

ilanschnell committed:
simplify logic in tree_traverse

ilanschnell committed:
deallocate tree when list creation fails

ilanschnell committed:
improve printed code table

ilanschnell committed:
add release date, update readme

ilanschnell committed:
Merge pull request #29 from DRMacIver/no_self_types

Fix types to actually be types

ilanschnell committed:
update changelog

ilanschnell committed:
update runall script

ilanschnell committed:
improve check for ambiguous prefix codes

ilanschnell committed:
update changelog

ilanschnell committed:
add comment

ilanschnell committed:
add comment

ilanschnell committed:
simplify decoding example

ilanschnell committed:
more consistent naming

ilanschnell committed:
update example readme

ilanschnell committed:
move check_codedict to C-level

ilanschnell committed:
remove Python level methods: encode, decode, iterdecode (in favor of having everthing on the C-level now

ilanschnell committed:
update changelog

ilanschnell committed:
add tests for non-iterable when encoding

ilanschnell committed:
update comment

ilanschnell committed:
wording

ilanschnell committed:
add test for undefined symbol in code

ilanschnell committed:
fix test for Python 2.6

ilanschnell committed:
typo

ilanschnell committed:
fix some tests for Python 2.5

ilanschnell committed:
add debug function to notes

ilanschnell committed:
formatting

ilanschnell committed:
formatting

ilanschnell committed:
update changelog

ilanschnell committed:
rework example to be much closer to C counterpart

ilanschnell committed:
better naming, add comment

ilanschnell committed:
add creation of .dot file to example

ilanschnell committed:
use different shape for leaf nodes

ilanschnell committed:
refactor Huffman example

ilanschnell committed:
ensure edges are always drawn, even from parent nodes with only one child

ilanschnell committed:
better naming

ilanschnell committed:
refactor Huffman example into library

ilanschnell committed:
add compress program

ilanschnell committed:
update work on Huffman example

ilanschnell committed:
inherit from object

ilanschnell committed:
add comments

ilanschnell committed:
add freq attribute again

ilanschnell committed:
update comment

ilanschnell committed:
add useful output

ilanschnell committed:
remove freq_string use collections.Counter instead

ilanschnell committed:
add count time to output

ilanschnell committed:
add encode time

ilanschnell committed:
add release date, update changelog and examples/README

ilanschnell committed:
update README.rst

ilanschnell committed:
fix fail to compile on Windows with VS 2015, issue #72

ilanschnell committed:
update changelog

ilanschnell committed:
prepare for 0.9.2 release

ilanschnell committed:
add better error message when trying to initialize from float or complex

ilanschnell committed:
update changelog

ilanschnell committed:
use macro to avoid duplicate code

ilanschnell committed:
fix typo, add comment, add assertion

ilanschnell committed:
refactor resize() - only shrink if the newsize falls lower than half the allocated size

ilanschnell committed:
update changelog

ilanschnell committed:
formatting

ilanschnell committed:
use sizeof(type)

ilanschnell committed:
add comment

ilanschnell committed:
fix #undef

ilanschnell committed:
add release date, update readme

ilanschnell committed:
cast newly allocated memory pointers

ilanschnell committed:
add comment

ilanschnell committed:
add test for reading 0 bytes from file

ilanschnell committed:
add comment to bitarray_fromfile()

ilanschnell committed:
add comment to bitarray_tofile()

ilanschnell committed:
update changelog

ilanschnell committed:
add comment to bitarray_frombytes()

ilanschnell committed:
add assert inf copy_n()

ilanschnell committed:
add comment to copy_n()

ilanschnell committed:
typo in changelog

ilanschnell committed:
add test to ensure repr is of type str

ilanschnell committed:
add a few tests

ilanschnell committed:
use PyBytes_* in C code, treating the Py3k function names as default (instead of PyString_*

ilanschnell committed:
enable unpack() to return either PyString or PyBytes object

ilanschnell committed:
add tests for creating bitarrys from unicode strings

ilanschnell committed:
fix bitarrays beings created from unicode in Python 2

ilanschnell committed:
only define PyBytes_* when necessary

ilanschnell committed:
unpack() to only return PyBytes

ilanschnell committed:
remove PyString_ redefinitions

ilanschnell committed:
next release will be 1.0.0, droping Python 2.5 support

ilanschnell committed:
remove obsolete comment

ilanschnell committed:
no reason to include bytesobject.h separately

ilanschnell committed:
remove special conditions for older Python versions from test suite

ilanschnell committed:
avoid import inside tests, as it may change the reference count

ilanschnell committed:
move remaining imports outside tests

ilanschnell committed:
update changelog

ilanschnell committed:
fix comment

ilanschnell committed:
consistently use str in doc-strings

ilanschnell committed:
fix wording for search() docstring

ilanschnell committed:
improve wording in comment

ilanschnell committed:
add tests for .index() method

ilanschnell committed:
add tests for .index() method

ilanschnell committed:
also search for 0 in arrys with mostly 1s

ilanschnell committed:
handle negative arguments of .index() method consistentlt with how they are treated for lists

ilanschnell committed:
update changelog

ilanschnell committed:
add another search test

ilanschnell committed:
update readme

ilanschnell committed:
update readme

ilanschnell committed:
wording and documentation updates

ilanschnell committed:
fix wording

ilanschnell committed:
typo

ilanschnell committed:
update readme

ilanschnell committed:
fix wording

ilanschnell committed:
add release date, update readme

ilanschnell committed:
typo

ilanschnell committed:
fix readme to pass twine check

ilanschnell committed:
add helper functions to exmaples

ilanschnell committed:
update documentation to use positional-only syntax in docstrings, and update README to pass Python 3 doctest

ilanschnell committed:
add tests for negative start and stop values in .test_index()

ilanschnell committed:
add normalize_index() to avoid duplicate code

ilanschnell committed:
add start and stop parameter to count(), and avoid calling setunused()

ilanschnell committed:
simplify test

ilanschnell committed:
DOC: improve formating for bitdiff

ilanschnell committed:
add optional start and stop parameters to .count() method

ilanschnell committed:
add error message when trying to run update-readme with Python version 2

ilanschnell committed:
add two new tests for new count() parameters

ilanschnell committed:
simplify example by using new funcionality

ilanschnell committed:
update changelog, next version will be 1.1.0

ilanschnell committed:
add official Python 3.8 support

ilanschnell committed:
improve ValueError excpetion message for bits2bytes when input is negative

ilanschnell committed:
simplify many tests by adding randombitarrays(start=1)
